Flaky test `StreamProcessorReplayTest.shouldReturnLastProcessedPositionInSnapshotWhenPausedDuringReplay`

**Summary**

> Try to answer the following as best as possible

- How often does the test fail? at least [once](https://github.com/camunda/zeebe/actions/runs/8343246949/job/22833094176)
- Does it block your work?
- Do we suspect that it is a real failure?

**Failures**

> Outline known failure cases, e.g. a failed assertion and its stacktrace obtained from Jenkins

<details><summary>Example assertion failure</summary>
<pre>
Warning:  Flakes: 
Warning:  io.camunda.zeebe.stream.impl.StreamProcessorReplayTest.shouldReturnLastProcessedPositionInSnapshotWhenPausedDuringReplay
Error:    Run 1: StreamProcessorReplayTest.shouldReturnLastProcessedPositionInSnapshotWhenPausedDuringReplay:319 » ConditionTimeout Condition with alias 'position has to be set when replay is paused' didn't complete within 10 seconds because assertion condition defined as a Lambda expression in io.camunda.zeebe.stream.impl.StreamProcessorReplayTest 
expected: 1L
 but was: -1L.
[INFO]   Run 2: PASS
</pre>
</details>

**Hypotheses**

> List any hypotheses if you have one; can be ommitted

**Logs**

> If possible, provide more context here, e.g. standard output logs, link to build, etc.

Full test results: [Google Drive](https://drive.google.com/file/d/1ejKbaMX8O2yDLX8G-IlKSmtD9bspr2Bn/view?usp=sharing)

<details><summary>Logs</summary>
<pre>

```
Error:  Tests run: 10, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 20.47 s <<< FAILURE! -- in io.camunda.zeebe.stream.impl.StreamProcessorReplayTest
Error:  io.camunda.zeebe.stream.impl.StreamProcessorReplayTest.shouldReturnLastProcessedPositionInSnapshotWhenPausedDuringReplay -- Time elapsed: 10.24 s <<< ERROR!
org.awaitility.core.ConditionTimeoutException: 
Condition with alias 'position has to be set when replay is paused' didn't complete within 10 seconds because assertion condition defined as a Lambda expression in io.camunda.zeebe.stream.impl.StreamProcessorReplayTest 
expected: 1L
 but was: -1L.
	at org.awaitility.core.ConditionAwaiter.await(ConditionAwaiter.java:167)
	at org.awaitility.core.AssertionCondition.await(AssertionCondition.java:119)
	at org.awaitility.core.AssertionCondition.await(AssertionCondition.java:31)
	at org.awaitility.core.ConditionFactory.until(ConditionFactory.java:1006)
	at org.awaitility.core.ConditionFactory.untilAsserted(ConditionFactory.java:790)
	at io.camunda.zeebe.stream.impl.StreamProcessorReplayTest.shouldReturnLastProcessedPositionInSnapshotWhenPausedDuringReplay(StreamProcessorReplayTest.java:319)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at java.base/java.util.concurrent.RecursiveAction.exec(RecursiveAction.java:194)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:387)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1312)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1843)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1808)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:188)
Caused by: org.opentest4j.AssertionFailedError: 
expected: 1L
 but was: -1L
	at java.base/jdk.internal.reflect.DirectConstructorHandleAccessor.newInstance(DirectConstructorHandleAccessor.java:62)
	at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:502)
	at io.camunda.zeebe.stream.impl.StreamProcessorReplayTest.lambda$shouldReturnLastProcessedPositionInSnapshotWhenPausedDuringReplay$8(StreamProcessorReplayTest.java:322)
	at org.awaitility.core.AssertionCondition.lambda$new$0(AssertionCondition.java:53)
	at org.awaitility.core.ConditionAwaiter$ConditionPoller.call(ConditionAwaiter.java:248)
	at org.awaitility.core.ConditionAwaiter$ConditionPoller.call(ConditionAwaiter.java:235)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)
```

</pre>
</details>

--- a/zeebe/stream-platform/src/test/java/io/camunda/zeebe/stream/impl/StreamProcessorReplayTest.java
+++ b/zeebe/stream-platform/src/test/java/io/camunda/zeebe/stream/impl/StreamProcessorReplayTest.java
@@ -295,8 +295,17 @@ final class StreamProcessorReplayTest {
             .processInstance(ELEMENT_ACTIVATING, Records.processInstance(1))
             .key(1)
             .causedBy((int) positionInSnapshot));
-    // starting the stream processor awaits until replay is completed
+
     streamPlatform.startStreamProcessor();
+    // Ensure that the last processed position is updated before taking a snapshot. This is
+    // necessary to ensure that after restarting the processor, it can recover from a non-zero
+    // processed position.
+    Awaitility.await("Last written position has to be updated before taking snapshot")
+        .untilAsserted(
+            () ->
+                assertThat(
+                        streamPlatform.getStreamProcessor().getLastProcessedPositionAsync().join())
+                    .isEqualTo(positionInSnapshot));
     streamPlatform.snapshot();
 
     // write more events to make sure that the processor is paused in replay phase after recovery

